compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -892,106 +886,158 @@ trait Implicits { self: Typer =>
+                    }
+                    instantiate(ctx.fresh.setExploreTyperState().setOwner(caseClass))
+                  case _ =>
+                    caseClass.typeRef
+                }
               }
             case child => child.termRef
           }
+
+          val (monoType, elemTypes) = mirroredType match {
+            case mirroredType: HKTypeLambda =>
+              val elems = mirroredType.derivedLambdaType(
+                resType = TypeOps.nestedPairs(cls.children.map(solve))
+              )
+              val AppliedType(tycon, _) = mirroredType.resultType
+              val monoType = AppliedType(tycon, mirroredType.paramInfos)
+              (monoType, elems)
+            case _ =>
+              val elems = TypeOps.nestedPairs(cls.children.map(solve))
+              (mirroredType, elems)
+          }
+
           val mirrorType =
-             mirrorCore(defn.Mirror_SumType, monoType, cls.name)
-              .refinedWith(tpnme.MirroredElemTypes, TypeAlias(TypeOps.nestedPairs(elemTypes)))
+             mirrorCore(defn.Mirror_SumType, monoType, mirroredType, cls.name)
+              .refinedWith(tpnme.MirroredElemTypes, TypeAlias(elemTypes))
+              .refinedWith(tpnme.MirroredElemLabels, TypeAlias(TypeOps.nestedPairs(elemLabels)))
           val mirrorRef =
-            if (cls.linkedClass.exists && !cls.is(Scala2x)) companionPath(monoType, span)
+            if (cls.linkedClass.exists && !cls.is(Scala2x)) companionPath(mirroredType, span)
             else anonymousMirror(monoType, ExtendsSumMirror, span)
           mirrorRef.cast(mirrorType)
-        case _ =>
-          EmptyTree
+        } else EmptyTree
       }
 
+      formal.member(tpnme.MirroredType).info match {
+        case TypeAlias(mirroredType) => mirrorFor(mirroredType)
+        case TypeBounds(mirroredType, _) => mirrorFor(mirroredType)
+        case _ => EmptyTree
+      }
+    }
+
   /** An implied instance for a type of the form `Mirror { type MirroredMonoType = T }`
    *  where `T` is a generic sum or product or singleton type.
    */
   lazy val synthesizedMirror: SpecialHandler =
-    (formal: Type, span: Span) => implicit (ctx: Context) =>
-      formal.member(tpnme.MirroredMonoType).info match {
-        case monoAlias @ TypeAlias(monoType) =>
-          if (monoType.termSymbol.is(CaseVal) || monoType.classSymbol.isGenericProduct)
-            synthesizedProductMirror(formal, span)(ctx)
-          else
-            synthesizedSumMirror(formal, span)(ctx)
+    (formal: Type, span: Span) => implicit (ctx: Context) => {
+      def mirrorFor(mirroredType: Type): Tree =
+        if (mirroredType.termSymbol.is(CaseVal) || mirroredType.classSymbol.isGenericProduct)
+          synthesizedProductMirror(formal, span)(ctx)
+        else
+          synthesizedSumMirror(formal, span)(ctx)
+
+      formal.member(tpnme.MirroredType).info match {
+        case TypeAlias(mirroredType) => mirrorFor(mirroredType)
```

Comment:
The `TypeAlias` case is redundant, since `TypeBounds` will match it anyway: a `TypeAlias(alias)` pattern also matches as `TypeBounds(alias, alias)`. This means we can keep the old logic without `mirrorFor`, only the pattern needs to be changed.

Reply:
Ahh ... right. That helps.